state/protocol/badger/state.go

```diff
@@ -41,7 +41,7 @@ type State struct {
 	params                      protocol.Params
 	protocolKVStoreSnapshotsDB  storage.ProtocolKVStore
 	epochProtocolStateEntriesDB storage.EpochProtocolStateEntries // TODO remove when EpochProtocolStateEntry is stored in KVStore
-	protocolState               protocol.MutableProtocolState
+	protocolState               protocol.ProtocolState
```

Comment:
Thanks for the context - very helpful. Also from my perspective, the current approach is totally fine. Essentially, `FollowerState` and `MutableProtocolState` just encapsulate a bunch of lower-level storage abstractions and implement higher-level logic on top, which is cleanly separated by our current design.

Reply:
A minor design oddity is the following cast from the interface to `badgerState.State` in all places where we instantiate a `FollowerState`:
https://github.com/onflow/flow-go/blob/557113c56be2deba08be5dbde260268a09275432/cmd/collection/main.go#L234-L247
I feel this is largely due to the fact that we use `badgerState.State` as a container for all the lower-level storage abstractions
*  Your code also uses it like that:  https://github.com/onflow/flow-go/blob/557113c56be2deba08be5dbde260268a09275432/state/protocol/badger/mutator.go#L79-L88
* I skimmed the code and it seems that the only place in `mutator.go` where we actually directly access a field in `badgerState.State` that is not a storage abstraction https://github.com/onflow/flow-go/blob/557113c56be2deba08be5dbde260268a09275432/state/protocol/badger/mutator.go#L792-L795 and this one could probably be encapsulated via a more refined interface (also improving clarity and modularity of the code)

I would be inclined to leave the code as is -- its only a minor aspect that could be improved in terms of clean abstractions. Nevertheless, just wanted to summarize my thoughts on how we could fix it if we wanted:
* Generally, the storage abstractions are initialized very early when booting a node. The storage abstractions (some subset of them) are injected into many many higher-level components. 
* If we had a container struct, that would bundle all the storage abstractions, we could:
  * more compactly represent the inputs for constructor methods: we input the container struct as opposed to each storage abstraction dependency separately (shorter list of method arguments for the constructors). The constructor just picks whatever they need
  * We don't need the concrete type (e.g. `badgerState.State`) just because we want to access the storage abstractions, as we do in the `FollowerState`.